src/ch9329.cpp

```diff
@@ -0,0 +1,401 @@
+{
+  auto byte_num = (static_cast<uint8_t>(p_key) >> 3) + 1;
+  auto bit_num = static_cast<uint8_t>(p_key) & 0b111;
+  m_data[byte_num] |= 1 << bit_num;
+  return *this;
+}
+
+ch9329::keyboard_media& ch9329::keyboard_media::release_media_key(
+  media_key p_key)
+{
+  auto byte_num = (static_cast<uint8_t>(p_key) >> 3) + 1;
+  auto bit_num = static_cast<uint8_t>(p_key) & 0b111;
+  m_data[byte_num] &= ~(1 << bit_num);
+  return *this;
+}
+
+ch9329::keyboard_media& ch9329::keyboard_media::release_all_keys()
+{
+  m_data = { 0x01, 0x00, 0x00, 0x00 };
+  return *this;
+}
+
+// keyboard acpi functions
+ch9329::keyboard_acpi& ch9329::keyboard_acpi::press_acpi_key(acpi_key p_key)
+{
+  auto bit_num = static_cast<uint8_t>(p_key);
+  m_data[1] |= 1 << bit_num;
+  return *this;
+}
+
+ch9329::keyboard_acpi& ch9329::keyboard_acpi::release_acpi_key(acpi_key p_key)
+{
+  auto bit_num = static_cast<uint8_t>(p_key);
+  m_data[1] &= ~(1 << bit_num);
+  return *this;
+}
+
+ch9329::keyboard_acpi& ch9329::keyboard_acpi::release_all_keys()
+{
+  m_data = { 0x02, 0x00 };
+  return *this;
+}
+
+// keyboard general functions
+ch9329::keyboard_general& ch9329::keyboard_general::press_control_key(
+  control_key_bit p_key)
+{
+  hal::byte mask = 0x00;
+  switch (p_key) {
+    case control_key_bit::left_control:
+      mask = 0b1;
+      break;
+    case control_key_bit::left_shift:
+      mask = 0b10;
+      break;
+    case control_key_bit::left_alt:
+      mask = 0b100;
+      break;
+    case control_key_bit::left_windows:
+      mask = 0b1000;
+      break;
+    case control_key_bit::right_control:
+      mask = 0b10000;
+      break;
+    case control_key_bit::right_shift:
+      mask = 0b100000;
+      break;
+    case control_key_bit::right_alt:
+      mask = 0b1000000;
+      break;
+    case control_key_bit::right_windows:
+      mask = 0b10000000;
+      break;
+  }
+  m_data[0] = m_data[0] | mask;
+  return *this;
+}
+
+ch9329::keyboard_general& ch9329::keyboard_general::release_control_key(
+  control_key_bit p_key)
+{
+  hal::byte mask = 0xFF;
+  switch (p_key) {
+    case control_key_bit::left_control:
+      mask = 0b11111110;
+      break;
+    case control_key_bit::left_shift:
+      mask = 0b11111101;
+      break;
+    case control_key_bit::left_alt:
+      mask = 0b11111011;
+      break;
+    case control_key_bit::left_windows:
+      mask = 0b11110111;
+      break;
+    case control_key_bit::right_control:
+      mask = 0b11101111;
+      break;
+    case control_key_bit::right_shift:
+      mask = 0b11011111;
+      break;
+    case control_key_bit::right_alt:
+      mask = 0b10111111;
+      break;
+    case control_key_bit::right_windows:
+      mask = 0b01111111;
+      break;
+  }
+  m_data[0] = m_data[0] & mask;
+  return *this;
+}
+
+ch9329::keyboard_general& ch9329::keyboard_general::press_normal_key(
+  normal_key p_key,
+  uint8_t p_slot)
+{
+  if (p_slot < 1) {
+    p_slot = 1;
+  } else if (p_slot > 6) {
+    p_slot = 6;
+  }
```

Comment:
```suggestion
  std::clamp(p_slot, 1, 6);
```

You can replace this with `std::clamp(p_slot, 1, 6)` see https://en.cppreference.com/w/cpp/algorithm/clamp.